src/ExtendedSystem.cpp

```diff
@@ -40,84 +40,171 @@ void ExtendedSystem::Init(PDBSetup &pdb, Velocity &vel,  config_setup::Input inp
+    for (p = placementStart, d = dataStart; p < placementEnd; ++p, ++d) {
+      pdb.atoms.x[p] = binaryCoor[d].x;
+      pdb.atoms.y[p] = binaryCoor[d].y;
+      pdb.atoms.z[p] = binaryCoor[d].z;
+    }
+  }
+}
 
-  for(; cmIndex < (int) molLookup.molLookupCount; cmIndex++) {
-    if(moleculeOffset >= numAtoms) break;
-    int currentMolecule = molLookup.molLookup[cmIndex];
-    int numberOfAtoms = mols.start[currentMolecule + 1] - mols.start[currentMolecule];
-    int atomDestinationStart = mols.start[currentMolecule];
+void ExtendedSystem::ReadCoordinate(PDBSetup &pdb, config_setup::Input inputFiles, MoleculeLookup & molLookup,
+                                     Molecules & mols){
+    // We must read restart PDB, which hold correct
+  // number atom info in each Box
+  numAtoms = 0;
+  boxStart[0] = 0;
+  for(int b = 0; b < BOX_TOTAL; b++) {
+    if(inputFiles.files.binaryCoorInput.defined[b]) {
+      if (mols.restartFromCheckpoint){
+        numAtomsInBox[b] = molLookup.restartedNumAtomsInBox[b];
+      } else {
+        numAtomsInBox[b] = pdb.atoms.numAtomsInBox[b];
+      }
+    }
+  }
 
-    for(int atom = 0; atom < numberOfAtoms; atom++) {
-      pdb.atoms.x[atomDestinationStart + atom] = binaryCoor[moleculeOffset + atom].x;
-      pdb.atoms.y[atomDestinationStart + atom] = binaryCoor[moleculeOffset + atom].y;
-      pdb.atoms.z[atomDestinationStart + atom] = binaryCoor[moleculeOffset + atom].z;
+  for(int b = 0; b < BOX_TOTAL; b++) {
+    if(inputFiles.files.binaryCoorInput.defined[b]) {
+      numAtoms += numAtomsInBox[b];
+      if (b == 1)
+        boxStart[1] = numAtomsInBox[0];
     }
+  }
 
-    moleculeOffset += numberOfAtoms;
+  binaryCoor.clear();
+  binaryCoor.resize(numAtoms);
+  for(int b = 0; b < BOX_TOTAL; b++) {
+    if(inputFiles.files.binaryCoorInput.defined[b]) {
+      std::string fName = inputFiles.files.binaryCoorInput.name[b];  
+      if (mols.restartFromCheckpoint)
+        read_binary_file(fName.c_str(), &binaryCoor[boxStart[b]], molLookup.restartedNumAtomsInBox[b]);
+      else
+        read_binary_file(fName.c_str(), &binaryCoor[boxStart[b]], pdb.atoms.numAtomsInBox[b]);
+    }
   }
+}
+
+void ExtendedSystem::UpdateMinMaxAtoms(PDBSetup &pdb, 
+                                      MoleculeLookup & molLookup,
+                                      Molecules & mols){
+  for (uint b = 0; b < BOX_TOTAL; b++) {
+    int stRange, endRange;
+    // -1 because we want to exclude the last array index  
+    // Box 0
+    // [0, numAtomsBox0)
+    // Box 1
+    // [numAtomsBox0, numAtomsBox0 + numAtomsBox1)
+
+    // To prevent segfault
+    if (numAtomsInBox[b] == 0)
+      return;
 
-  delete [] binaryCoor;
+    if (b == 0){
+      stRange = 0;
+      endRange = numAtomsInBox[0] - 1;
+    } else if (b == 1) {
+      stRange = numAtomsInBox[0];
+      endRange = stRange + numAtomsInBox[1] - 1;
+    } else {
+      std::cout << "Error: Only Box 0 and Box 1 supported!" << std::endl;
+      exit(EXIT_FAILURE);
+    }
```

Comment:
Can't you make it more general to support simulation with more than 2 boxes?

Reply:
Done